visualizer/draw/bubbleprof-ui.js

```diff
@@ -568,24 +590,23 @@ class BubbleprofUI extends EventEmitter {
 
     if (window.location.hash) {
       setTimeout(() => {
-        const id = parseInt(window.location.hash.slice(2))
-        switch (window.location.hash.charAt(1)) {
-          case 'a':
-            const aggregateNode = this.dataSet.aggregateNodes.get(id)
-            const uiWithinAggregate = this.jumpToNode(aggregateNode)
-            this.emit('navigation', { from: this, to: uiWithinAggregate })
-            break
-          case 'c':
-            const clusterNode = this.dataSet.clusterNodes.get(id)
-            const uiWithinCluster = this.jumpToNode(clusterNode)
-            this.emit('navigation', { from: this, to: uiWithinCluster })
-            break
-          case 'x':
-            this.parseCollapsedNodeHash(window.location.hash)
-            break
-        }
+        this.parseHash(window.location.hash.slice(1))
       })
     }
+
+    window.addEventListener('popstate', (event) => {
+      const { hash } = event.state
+      if (this.topMostUI) {
+        this.topMostUI.traverseUp(null, { silent: true })
+        // Close stack frames when moving away from their node.
+        if (this.topMostUI.selectedDataNode) {
+          this.topMostUI.clearFrames()
+        }
```

Comment:
We've got a bug here - it doesn't handle sublayouts within sublayouts (ie. cluster nodes within collapsed nodes, or collapsed nodes within collapsed nodes). 

I think this slipped through your manual testing because by chance, you managed to generate a profile with no collapsed nodes! This is pretty rare, I'd be interested to see the application you used to generate it and add this sample to the sample set, it's pretty unique :-)

Here's an example that demonstrates the bug created using this PR: https://upload.clinicjs.org/public/07396a39c703c9ba755914d9e909033778049297e754f7f2767ff7b9b3d7cc66/11082.clinic-bubbleprof.html - click into the main MongoDB bubble, then click in to the first node in the chain (name ends with "&..."), then use the back button.

In trying to understand why the bug happens, I think I found something that fixes it, but I've only done very basic testing on this code change, so experiment with it, it probably needs altering: 

```
    window.addEventListener('popstate', (event) => {
      const { hash } = event.state
      // Close stack frames when moving away from their node.
      if (this.topMostUI.selectedDataNode) {
        this.topMostUI.clearFrames()
      }

      if (hash) {
        this.topMostUI.parseHash(hash)
      } else {
        this.topMostUI.traverseUp(null, { silent: true })
      }
    })
```

Sample: https://upload.clinicjs.org/public/1324c29b45787850e89d4c81c7de223379c49f4fad86a94a94dc802ccff023d6/11082.clinic-bubbleprof.html

Changes explained:

- If we always use traverseUp, we're assuming we need to go to all the way to the main layout first, which isn't true if we're stepping back within nested sublayouts. The jumpToNode / selectNode chain called from parseHash should handle any steps back up the tree required, it's designed to handle and animate any jump from anywhere to anywhere (e.g. using the buttons in the stack frames or search feature), so we shouldn't need to worry about that here unless we know we're going _all_ the way back because it's an empty hash. 
- `parseHash` should be called from the topmostUI here. Jumps to collapsed nodes need awareness of the UI and lineage of parent UIs. This is a tricky thing to understand, I discussed it briefly earlier but it takes some time and some experimentation with practical examples to get used to it. For example, if there is a hash `#x2-x1-c2` (x = collapsed node, c = cluster node), that means "find x2 within x1 within c2"; c2 (cluster node 2) can be located from anywhere because cluster nodes are unique and entirely determined by the clusterNode data from the dataSet, but the x1 referred to in the hash can only be located from within the context of c2, because it was only while processing the c2 cluster node's sublayout that it was identified that some nodes should be collapsed together to form collapsed node x1; and the x2 in the hash can only be located from the context of the x1 within c2 for the same reason. So, if were are in `#x1-c2` and we were previously in `#x2-x1-c2`, handling a back button press needs to look for `x2` using the context of the UI and layout objects of `#x1-c2`.

Reply:
Ahh, doing `topMostUI.parseHash()` instead of `originalUI.parseHash()` also fixes the problem I encountered earlier, where moving to a different layout using these arrows would cause both layouts to show at once:
![image](https://user-images.githubusercontent.com/1006268/44903145-0c289880-ad0c-11e8-98d6-368076c64ebb.png)
Calling `traverseUp()` first fixed that as well but as you point out that's not quite correct.

With only this change, moving back from a hash like `#x1-x1-c3` to `#x1-c3` shows both the x1-c3 and c3 layouts at the same time. Time to test if my understanding of the layouts is now sufficient to debug this :)

Reply:
Ok, using an animationQueue array in `parseCollapsedHash` fixed it!